File: apps/authentication/src/service/AuthenticationService.ts
'use strict'

import {AuthenticationDao} from "../dao/AuthenticationDao";
import {RegisterRequest} from "../model/request/RegisterRequest";
import {LoginRequest} from "../model/request/LoginRequest";
import {ConfirmResponse, LoginResponse, LogoutResponse, RegisterResponse} from "../model/response/Response";
import {InvalidRequestError, UserNotFoundError, UserEmailAlreadyExistsError} from "../error/Errors";
import {LogoutRequest} from "../model/request/LogoutRequest";
import {AdminGetUserResponse, SignUpResponse, UserType} from "aws-sdk/clients/cognitoidentityserviceprovider";
import {ConfirmSignupRequest} from "../model/request/ConfirmSignupRequest";
import {ConfirmCodeRequest} from "../model/request/ConfirmCodeRequest";
import {SendCodeRequest} from "../model/request/SendCodeRequest";
import {User} from "../model/User";
import {UpdateRequest} from "../model/request/UpdateRequest";
import {ForgotPasswordRequest} from "../model/request/ForgotPasswordRequest";
import {ResetPasswordRequest} from "../model/request/ResetPasswordRequest";

class AuthenticationService {
    private authenticationDao: AuthenticationDao;
    constructor() {
        this.authenticationDao = new AuthenticationDao()
    }

    async getUser(userId: string): Promise<User> {
        const response = await this.authenticationDao.getUser(userId)
        return this._mapToUser(response)
    }

    async getUserByPhone(phoneNumber: string): Promise<User | null> {
        const response = await this.authenticationDao.getUserByPhone(phoneNumber)
        if (!response) {
            return null
        }

        return this._mapToUserFromUserType(response)
    }

    async register(registerRequest: RegisterRequest): Promise<RegisterResponse> {
        // create authentication account
        const signUpResponse: SignUpResponse = await this.authenticationDao.register(registerRequest)

        return new RegisterResponse(signUpResponse)
    }

    async confirm(confirmRequest: ConfirmSignupRequest): Promise<ConfirmResponse> {
        return await this.authenticationDao.confirm(confirmRequest)
    }

    async login(loginRequest: LoginRequest): Promise<LoginResponse> {
        const authResult = await this.authenticationDao.login(loginRequest)

        if (authResult === undefined) {
            throw new InvalidRequestError("Received undefined login response")
        }

        let user = await this.authenticationDao.getUserByPhone(loginRequest.username)
        if (!user) {
            user = await this.authenticationDao.getUserByEmail(loginRequest.username)
        }

        return new LoginResponse(user?.Username!, authResult.AuthenticationResult!);
    }

    async logout(logoutRequest: LogoutRequest): Promise<LogoutResponse> {
        await this.authenticationDao.logout(logoutRequest)
        return new LogoutResponse()
    }

    async forgotPassword(request: ForgotPasswordRequest) {
        const user = await this.authenticationDao.getUserByPhone(request.username)
        if (!user) {
            return
        }

        await this.authenticationDao.forgotPassword(user!.Username!)
    }

    async confirmForgotPassword(request: ResetPasswordRequest) {
        const user = await this.authenticationDao.getUserByPhone(request.username)
        if (!user) {
            throw new UserNotFoundError()
        }

        await this.authenticationDao.confirmForgotPassword(user!.Username!, request.code, request.password)
    }

    async sendCode(request: SendCodeRequest) {
        return await this.authenticationDao.sendVerificationCode(request)
    }

    async confirmCode(request: ConfirmCodeRequest): Promise<string> {
        await this.authenticationDao.verifyCode(request)
        const user = await this.authenticationDao.getUserByToken(request.accessToken)
        return user?.Username
    }

    async updateUser(userId: string, request: UpdateRequest): Promise<User> {
        // Only one email can exist per user
        if (request.email !== undefined) {
            const user = await this.authenticationDao.getUserByEmail(request.email)
            if (user && this._mapToUserFromUserType(user).userId !== userId) {
                throw new UserEmailAlreadyExistsError(request.email)
            }
        }

        await this.authenticationDao.updateUser(userId, request)
        return await this.getUser(userId)
    }

    _mapToUser(response: AdminGetUserResponse): User {
        const attributes = response.UserAttributes!

        // always present after registration
        const userId = attributes.find(a => a.Name === 'sub')!.Value!
        const phone = attributes?.find(a => a.Name === 'phone_number')?.Value!
        const phoneVerified = attributes?.find(a => a.Name === 'phone_number_verified')?.Value === 'true'

        // not always present
        const email = attributes.find(a => a.Name === 'email')!.Value
        const emailVerified = attributes.find(a => a.Name === 'email_verified')!.Value === 'true'
        const name = attributes?.find(a => a.Name === 'name')?.Value

        return new User(userId, phone, phoneVerified, email, emailVerified, name)
    }

    _mapToUserFromUserType(response: UserType): User {
        const attributes = response.Attributes!

        // always present after registration
        const userId = attributes.find(a => a.Name === 'sub')!.Value!
        const phone = attributes?.find(a => a.Name === 'phone_number')!.Value!
        const phoneVerified = attributes?.find(a => a.Name === 'phone_number_verified')!.Value! === 'true'

        // not always present
        const email = attributes.find(a => a.Name === 'email')?.Value
        const emailVerified = attributes.find(a => a.Name === 'email_verified')?.Value === 'true'
        const name = attributes?.find(a => a.Name === 'name')?.Value

        return new User(userId, phone, phoneVerified, email, emailVerified, name)
    }
}

export {AuthenticationService}

File: apps/authentication/src/service/authentication_service.py
import logging
from typing import Optional, List

from types_boto3_cognito_idp.type_defs import AttributeTypeTypeDef, GetUserAttributeVerificationCodeResponseTypeDef

from src.dao.authentication_dao import AuthenticationDao
from src.model.errors import InvalidRequestError, UserNotFoundError, UserEmailAlreadyExistsError, \
    PasswordResetRequiredError
from src.model.requests import RegisterRequest, ConfirmSignupRequest, LoginRequest, RefreshLoginRequest, \
    LogoutRequest, ResendSignupCodeRequest, ForgotPasswordRequest, ResetPasswordRequest, SendCodeRequest, \
    ConfirmCodeRequest, UpdateRequest
from src.model.response import LoginResponse, RegisterResponse, RefreshLoginResponse, ForgotPasswordResponse
from src.model.user import User

logger = logging.getLogger(__name__)


def _map_to_user(attributes: List[AttributeTypeTypeDef]) -> User:
    user_id: str = next((a for a in attributes if a["Name"] == "sub"))["Value"]
    phone_number: str = next((a for a in attributes if a["Name"] == "phone_number"))["Value"]
    phone_number_verified: bool = bool(next((a for a in attributes if a["Name"] == "phone_number_verified"))["Value"])

    email_attribute = next((a for a in attributes if a["Name"] == "email"), None)
    email_verified_attribute = next((a for a in attributes if a["Name"] == "email_verified"), None)
    name_attribute = next((a for a in attributes if a["Name"] == "name"), None)

    return User(
        user_id=user_id,
        phone_number=phone_number,
        phone_number_verified=phone_number_verified,
        email=email_attribute["Value"] if email_attribute is not None else None,
        email_verified=bool(email_verified_attribute["Value"]) if email_verified_attribute is not None else None,
        name=name_attribute["Value"] if name_attribute is not None else None
    )


class AuthenticationService:
    def __init__(self):
        self.authentication_dao = AuthenticationDao()

    def get_user(self, user_id: str) -> User:
        res = self.authentication_dao.get_user(user_id=user_id)
        return _map_to_user(res.get("UserAttributes", []))

    def get_user_by_phone(self, phone_number: str) -> Optional[User]:
        res = self.authentication_dao.get_user_by_phone(phone_number=phone_number)
        if not res:
            return None

        return _map_to_user(res.get("Attributes", []))

    def register(self, request: RegisterRequest) -> RegisterResponse:
        res = self.authentication_dao.register(username=request.username, password=request.password)
        return RegisterResponse(res)

    def confirm(self, request: ConfirmSignupRequest):
        self.authentication_dao.confirm(username=request.username, confirmation_code=request.confirmation_code)

    def login(self, request: LoginRequest) -> LoginResponse:
        auth_result = self.authentication_dao.login(username=request.username, password=request.password)
        if not auth_result:
            raise InvalidRequestError("Failed to login")

        if auth_result.get("ChallengeName") == "NEW_PASSWORD_REQUIRED":
            raise PasswordResetRequiredError()

        user = self.authentication_dao.get_user_by_phone(phone_number=request.username)
        if not user:
            print(f"Failed to get user by phone, trying by email instead")
            user = self.authentication_dao.get_user_by_email(email=request.username)

        return LoginResponse(
            id=user.get("Username"),
            auth=auth_result.get("AuthenticationResult")
        )

    def refresh_login(self, request: RefreshLoginRequest) -> RefreshLoginResponse:
        auth_result = self.authentication_dao.refresh_login(refresh_token=request.refresh_token)
        if not auth_result:
            raise InvalidRequestError("Failed to refresh login")

        return RefreshLoginResponse(auth=auth_result.get("AuthenticationResult"))

    def logout(self, request: LogoutRequest):
        self.authentication_dao.logout(access_token=request.access_token)

    def resend_signup_code(self, request: ResendSignupCodeRequest):
        user = self.authentication_dao.get_user_by_phone(phone_number=request.username)
        if not user or user.get("UserStatus") != "UNCONFIRMED":
            print(f"User does not exist or is not unconfirmed")
            return

        self.authentication_dao.resend_confirmation_code(username=request.username)

    def forgot_password(self, request: ForgotPasswordRequest) -> Optional[ForgotPasswordResponse]:
        user = self.authentication_dao.get_user_by_phone(phone_number=request.username)
        if not user:
            return None

        res = self.authentication_dao.forgot_password(username=request.username)
        return ForgotPasswordResponse(res)

    def confirm_forgot_password(self, request: ResetPasswordRequest):
        user = self.authentication_dao.get_user_by_phone(phone_number=request.username)
        if not user:
            raise UserNotFoundError()

        self.authentication_dao.confirm_forgot_password(username=request.username,
                                                        password=request.password,
                                                        code=request.code)

    def send_code(self, request: SendCodeRequest) -> GetUserAttributeVerificationCodeResponseTypeDef:
        return self.authentication_dao.send_verification_code(access_token=request.access_token,
                                                              attribute_name=request.attribute_name)

    def confirm_code(self, request: ConfirmCodeRequest) -> str:
        self.authentication_dao.verify_code(access_token=request.access_token,
                                            attribute_name=request.attribute_name,
                                            code=request.code)
        user = self.authentication_dao.get_user_by_token(access_token=request.access_token)
        return user.get("Username")

    def update_user(self, user_id: str, request: UpdateRequest) -> User:
        # Only one email can exist per user
        if request.email:
            user = self.authentication_dao.get_user_by_email(email=request.email)
            if user and _map_to_user(user.get("Attributes")).user_id != user_id:
                raise UserEmailAlreadyExistsError(request.email)

        self.authentication_dao.update_user(username=user_id, email=request.email)
        return self.get_user(user_id=user_id)
